### PR TITLE
Adjust rv of C_GetSlotList and C_GetMechanismList for NULL buffer.

### PR DESCRIPTION
**Pull request by EDVTAZ**
Relevant documentation is on page 72 and 76 of the pkcs11 base v2.40 specification.

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

### CODE REVIEW

File: libsks/src/pkcs11_token.c [hunk 1]
```diff
@@ -208,6 +209,10 @@ CK_RV sks_ck_token_mechanism_ids(CK_SLOT_ID slot,
 			  &ctrl, sizeof(ctrl), NULL, 0, outbuf, &outsize);
 	if (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL)
 		*count = outsize / sizeof(uint32_t);
+	if(!mechanisms && rv == CKR_BUFFER_TOO_SMALL) {
+		rv = CKR_OK;
+		goto bail;
+	}
```
Comment (etienne-lms) [inline libsks/src/pkcs11_token.c]: I think the code at line 204 should also be changed. `*count` maybe meaningless if caller queries the number of available mechanisms.
Should be something like:
```c
	rv = ck_invoke_ta(NULL, SKS_CMD_CK_MECHANISM_IDS,
			  &ctrl, sizeof(ctrl), NULL, 0, NULL, &outsize);

	if (rv != CKR_BUFFER_TOO_SMALL)
		return CKR_FUNCTION_FAILED;

	if (!mechanisms) {
		*count = outsize / sizeof(uint32_t);
		return CKR_OK;
	}
	if (*count < outsize / sizeof(uint32_t)) {
		*count = outsize / sizeof(uint32_t);
		return rv;
	}

	/* Get the mechanisms IDs now */
	outbuf = malloc(outsize);
 	if (!outbuf)
 		return CKR_HOST_MEMORY;

	rv = ck_invoke_ta(NULL, SKS_CMD_CK_MECHANISM_IDS,
 			  &ctrl, sizeof(ctrl), NULL, 0, outbuf, &outsize);

	if (rv == CKR_OK) {
		if (sks2ck_mechanism_type_list(mechanisms, outbuf, *count)) {
			LOG_ERROR("unexpected bad mechanism_type list\n");
			rv = CKR_DEVICE_ERROR;
		}
	}

	free(outbuf);
	return rv;
```

File: libsks/src/pkcs11_token.c [hunk 2]
```diff
@@ -66,6 +66,7 @@ CK_RV sks_ck_slot_get_list(CK_BBOOL present,
 
 	if (!slots || *count < (size / sizeof(uint32_t))) {
 		*count = size / sizeof(uint32_t);
+		if(!slots) return CKR_OK;
```
Comment (etienne-lms) [inline libsks/src/pkcs11_token.c]: Nice, thanks.
Can you expand over several line, for consistency of the code layout:
```
if (!slot || *count < (size / sizeof(uint32_t))) {
    *count = size / sizeof(uint32_t);
    if (!slots)
        return CKR_OK;
   return CKR_BUFFER_TOO_SMALL;
}
```

**Comment by etienne-lms** [comment]
Looks ok, thanks Gábor,
Can you squash the 2 commits into a single one with your singed-of-by inside? i will (test and) merge.

**Comment by EDVTAZ** [comment]
Is it all right like this?

File: libsks/src/pkcs11_token.c [hunk 3]
```diff
@@ -66,6 +66,8 @@ CK_RV sks_ck_slot_get_list(CK_BBOOL present,
 
 	if (!slots || *count < (size / sizeof(uint32_t))) {
 		*count = size / sizeof(uint32_t);
+		if(!slots)
```
Comment (etienne-lms) [inline libsks/src/pkcs11_token.c]: minor: add a space before the parenthesis.

**Review by etienne-lms — COMMENTED** [review]
See the minor comment.
Also, for the commit comment, keep only the useful part, i.e:
```
libsks: Adjust rv of C_GetSlotList and C_GetMechanismList for NULL buffer.

Functions shall return CKR_OK when caller only queries the number of slots and
number of mechanisms.
Relevant documentation is on page 72 and 76 of the pkcs11 base v2.40 specification.

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>
```

**Comment by EDVTAZ** [comment]
Is it good like this? Should I delete all the previous commit comments from the final squashed or should I leave it like this?

**Comment by etienne-lms** [comment]
Thanks @EDVTAZ, i have reformat the commit comment so it looks better and added some parentheses to clarify the implementation. 

Closing, merge manually.
Related update on optee_test (https://github.com/etienne-lms/optee_test/pull/6) merged.

Thanks for the contribution.

**Pull request closed without merging**